File: leet_2352_equal_pairs.go
func equalPairs(grid [][]int) int {
    dict := make(map[string]int)
	len := len(grid)
	cols := make([]string, len)
	res := 0

	for _, nums := range grid {
		s := ""
		for i := 0; i < len; i++ {
			n := nums[i]
			s += " " + fmt.Sprint(n)
			cols[i] += " " + fmt.Sprint(n)
		}

		if _, ok := dict[s]; ok {
			dict[s]++
		} else {
			dict[s] = 1
		}
		// fmt.Println(s, dict[s])
	}

	for _, c := range cols {
		if _, ok := dict[c]; ok {
			res += dict[c]
		}
	}
    return res
}

File: leet_2352_equal_pairs.py
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        D = defaultdict(lambda: 0)
        n = len(grid)
        cols = ['' for _ in range(n)]
        for line in grid:
            s = ' '.join([str(_) for _ in line])
            for i in range(n):
                cols[i] += ' ' + str(line[i])
            D[s] += 1
        res = 0
        for c in cols:
            c = c[1:]
            res += D[c]
        return res
